ExampleMod Example for Projectiles Moving in a Sine Wave

absoluteAquarian opened the pull request:
### What are the changes to ExampleMod?
- added `Content/Projectiles/ExampleSineWaveMotionProjectile.cs`, which serves as an example of making projectiles move in a sine wave
- added `Content/Items/Weapons/ExampleMagicRod.cs`, the item that spawns the `ExampleSineWaveMotionProjectile` projectile

### Do these changes need additional documentation?
Not sure

absoluteAquarian committed:
added example projectile for sine wave movement

absoluteAquarian committed:
more accurate variable name

direwolf420 commented:
Overall agree that this should be added. There's even guides on the forums for such a thing: https://forums.terraria.org/index.php?threads/creating-a-projectile-that-moves-in-a-sive-wave-pattern.83245/#post-1821219

One small formatting thing I noticed:
```
} else {
```
should be split like

```
}
else {
```

absoluteAquarian committed:
formatting

absoluteAquarian committed:
forgot to change the formatting for an else statement

Chicken-Bones commented on ExampleMod/Content/Projectiles/ExampleSineWaveMotionProjectile.cs:
```diff
@@ -0,0 +1,150 @@
+			float wavesPerSecond = 2;
+
+			// Math.Sin expects a radians angle instead of degrees
+			float radians = MathHelper.ToRadians(sineTimer * 6f * wavesPerSecond);
```
Okay, sure, 'ticks' * 6 gives you 1 revolution per second, but uhh, not very clear. Instead I suggest:
```cs
float waveProgress = sineTimer / 60f * wavesPerSecond; // 1 for each full sine wave
float angle = waveProgress * 2 * MathF.PI; // angles in radians
```

Chicken-Bones commented on ExampleMod/Content/Projectiles/ExampleSineWaveMotionProjectile.cs:
```diff
@@ -0,0 +1,150 @@
+
+			// Math.Sin expects a radians angle instead of degrees
+			float radians = MathHelper.ToRadians(sineTimer * 6f * wavesPerSecond);
+			float sine = (float)Math.Sin(radians) * trueDirection;
```
`MathF.Sin` exists

Chicken-Bones commented on ExampleMod/Content/Projectiles/ExampleSineWaveMotionProjectile.cs:
```diff
@@ -0,0 +1,150 @@
+			// Update the rotation used to draw the projectile
+			// This projectile should act as if it were moving along the sine wave
+			float cosine = (float)Math.Cos(radians) * trueDirection;
+			Projectile.rotation = Projectile.velocity.ToRotation() + MathHelper.PiOver4 * cosine * -1;
```
Why add the `cosine` here? Not clear how this math works.

absoluteAquarian replied:
The derivative of `sin(x)` is `-cos(x)`.  I use this derivative to make the projectile "point" toward where it will be moving.

Chicken-Bones commented on ExampleMod/Content/Projectiles/ExampleSineWaveMotionProjectile.cs:
```diff
@@ -0,0 +1,150 @@
+
+		public override void AI() {
+			// Ensure that the "waveDirection = 1" projectile starts above the "waveDirection = -1" projectile not matter which direction they were fired from
+			int trueDirection = Projectile.direction * waveDirection;
```
Why use `Projectile.direction` here? I don't see where it's set.

absoluteAquarian replied:
Isn't it set automatically?

absoluteAquarian replied:
I double checked, and it is.

absoluteAquarian replied:
nevermind on the previous two comments.  `Projectile.direction` is set later in AI

Chicken-Bones commented on ExampleMod/Content/Items/Weapons/ExampleMagicRod.cs:
```diff
@@ -0,0 +1,72 @@
+				Projectile projectile = Projectile.NewProjectileDirect(source, position, velocity, type, damage, knockback, player.whoAmI);
+
+				ExampleSineWaveMotionProjectile modProjectile = projectile.ModProjectile as ExampleSineWaveMotionProjectile;
+				modProjectile.waveDirection = waveDirection;
```
Suggest that instead of using `waveDirection` you use `waveOffset` (from 0-1, added to angle before multiplying by 2pi) that way if you had 4 projectiles you could do fractions.

Chicken-Bones commented on ExampleMod/Content/Projectiles/ExampleSineWaveMotionProjectile.cs:
```diff
@@ -0,0 +1,150 @@
+			Projectile.rotation = Projectile.velocity.ToRotation() + MathHelper.PiOver4 * cosine * -1;
+
+			// Update the frame used to draw the projectile
+			const float cosineOf45Degrees = 0.707106781f;
+			float cosineAbsolute = Math.Abs(cosine);
+			if (sine > 0) {
+				// Offset is counter-clockwise to the direction
+				if (cosineAbsolute > cosineOf45Degrees) {
+					Projectile.frame = Projectile.direction == 1 ? 0 : 2;
+				}
+				else {
+					Projectile.frame = 1;
+				}
+			}
+			else if (sine < 0) {
+				// Offset is clockwise to the direction
+				if (cosineAbsolute > cosineOf45Degrees) {
+					Projectile.frame = Projectile.direction == 1 ? 2 : 0;
+				}
+				else {
+					Projectile.frame = 1;
+				}
+			}
+			else {
+				// Offset is zero
+				Projectile.frame = 1;
+			}
```
Comments are needed to explain more about the relationship between sprite and math here. My understanding is that you just want to change the sprite when you're at a 'peak' or 'trough' to indicate it'll be turning around.

You can detect those just by comparing `sine` though, and if I'm reading the math correctly, you currently have the 'curved' sprites (0 and 2) used when the projectile is travelling straight (0-45, 135-225, 315-... degrees)

absoluteAquarian replied:
changed to only check `sine`

Chicken-Bones commented on ExampleMod/Content/Projectiles/ExampleSineWaveMotionProjectile.cs:
```diff
@@ -0,0 +1,150 @@
+			sineTimer++;
+		}
+
+		public override bool PreDraw(ref Color lightColor) {
```
I'm pretty sure this shouldn't be required. Why is positioning and framing the texture not enough? By altering the draw here, you can create differences between draw and hitbox.

absoluteAquarian replied:
i tried to do that via the DrawOffset variables, but it was annoying to make them work when the projectile was reversed

absoluteAquarian committed:
cleanup and various changes

- made ExampleMagicRod spawn 4 projectiles instead of 2
- made the Shoot code more friendly to varying projectile counts
- moved the `Projectile.direction` assignment to earlier in the AI
- simplified the code reponsible for setting `Projectile.frame`
- changed `waveDirection` to `waveOffset` to be more friendly to custom wave offsets (instead of just changing the parity)
- added comments explaining why `Projectile.rotation` is set the way it is
- added code to make the wave amplitude grow to the intended size (so that all of the projectiles spawn from the player's center)
- changed the code reponsible for setting the wave offset to be more intuitive

absoluteAquarian committed:
fixed comment typo

absoluteAquarian committed:
updated outdated comment

absoluteAquarian committed:
fixed the rotation calculations

absoluteAquarian committed:
added explanation for why -1 is multiplied to the slope for `Math.Atan()`

absoluteAquarian committed:
updated comment

absoluteAquarian committed:
changed the projectile code to be based on "wave stride" instead of "waves per second" + other smaller adjustments

absoluteAquarian committed:
skip movement code if velocity is zero, since the projectile couldn't move anyway
